src/Parameter.cpp

```diff
@@ -71,80 +72,66 @@ void Parameter::check_dim_ok(int dim) const {
         << "Dimension " << dim << " is not in the range [0, " << dimensions() - 1 << "]\n";
 }
 
+void Parameter::register_if_needed() {
+    if (contents.defined() && contents->should_be_registered) {
+        ObjectInstanceRegistry::register_instance(this, 0, ObjectInstanceRegistry::FilterParam, this, nullptr);
+    }
+}
+
+void Parameter::unregister_if_needed() {
+    if (contents.defined() && contents->should_be_registered) {
+        // This will assert-fail if not registered.
+        ObjectInstanceRegistry::unregister_instance(this);
+    }
+}
+
 Parameter::Parameter() : contents(nullptr) {
     // Undefined Parameters are never registered.
 }
 
 Parameter::Parameter(Type t, bool is_buffer, int d) :
-    contents(new ParameterContents(t, is_buffer, d, unique_name('p'), false, true)) {
+    contents(new ParameterContents(t, is_buffer, d, unique_name('p'), false, /* should_be_registered */ true)) {
     internal_assert(is_buffer || d == 0) << "Scalar parameters should be zero-dimensional";
-    // Note that is_registered is always true here; this is just using a parallel code structure for clarity.
-    if (contents.defined() && contents->is_registered) {
-        ObjectInstanceRegistry::register_instance(this, 0, ObjectInstanceRegistry::FilterParam, this, nullptr);
-    }
+    register_if_needed();
 }
 
 Parameter::Parameter(Type t, bool is_buffer, int d, const std::string &name, bool is_explicit_name, bool register_instance) :
     contents(new ParameterContents(t, is_buffer, d, name, is_explicit_name, register_instance)) {
     internal_assert(is_buffer || d == 0) << "Scalar parameters should be zero-dimensional";
-    if (contents.defined() && contents->is_registered) {
-        ObjectInstanceRegistry::register_instance(this, 0, ObjectInstanceRegistry::FilterParam, this, nullptr);
-    }
+    register_if_needed();
 }
 
 Parameter::Parameter(const Parameter& that) : contents(that.contents) {
-    if (contents.defined() && contents->is_registered) {
-        ObjectInstanceRegistry::register_instance(this, 0, ObjectInstanceRegistry::FilterParam, this, nullptr);
-    }
+    register_if_needed();
```

Comment:
I doubt we actually need to register copies of Parameter objects, except for maybe the case where you're copying a Generator object. Are they even copyable?

Reply:
The problem is that there is this idiom in existing Generators for specifying an ImageParams type at build time:

     GeneratorParam<Type> some_type_{...}
     ImageParam input_{UInt(8), 3, "input"};
     .... 
     Func build() { input_ = ImageParam(some_type, 3, "input"); }

In this case, the ImageParam itself doesn't change, but its contents do.

(There is a cleaner way to do this with Input<>, and all Input<> and Output<> are noncopyable and nonassignable)